Ev/cdr bug fix

eyalvo-matrix opened the pull request:
Hi to all , 
Reviewing the CDR engine  I have noticed that in case of a form that is being created and deleted 
the CDR still consider the form as created.
I added a code to fix that. 
The code apply only to standard forms. 
Assuming "form_ " + name format .

mdsupport commented:
[PR 359](https://github.com/openemr/openemr/pull/359/commits/5a9aac9fa6ad59800133ae064898b3cddb96f5ab) provides a generic approach to configuring "non standard" tables - those without "form_" prefix, "id" as primary key or 1:1 records. 

Until an encounter object is put in place or project starts using sql views, bugs will rule the codebase.

eyalvo-matrix commented:
Hi fellow developer,

Thank you for your replay.
I fear I did not fully understand your answer .
My fix is indeed specific for form_  prefix .
I understand that in the future or even now forms can be configured in a
different way.
We currently do need this fix since our OpenEmr is already in use .
I attached an example of our Db
Can you please tell me what approach should we take to solve the problem?
1:1 id is currently not an option for my best understanding.

thanks in advance ,

Eyal



On Sun, Nov 27, 2016 at 8:54 PM, mdsupport <notifications@github.com> wrote:

> PR 359
> <https://github.com/openemr/openemr/pull/359/commits/5a9aac9fa6ad59800133ae064898b3cddb96f5ab>
> provides a generic approach to configuring "non standard" tables - those
> without "form_" prefix, "id" as primary key or 1:1 records.
>
> Until an encounter object is put in place or project starts using sql
> views, bugs will rule the codebase.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/openemr/openemr/pull/360#issuecomment-263139614>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ATqohPfMVXdp7r4RVjksKLTi7m0DE5o9ks5rCdH0gaJpZM4K9LPB>
> .
>

bradymiller commented on library/clinical_rules.php:
```diff
@@ -1914,13 +1914,28 @@ function exist_database_item($patient_id,$table,$column='',$data_comp,$data='',$
-      "WHERE `" . add_escape_custom($column) ."`" . $compSql . "? " .
-      "AND " . add_escape_custom($patient_id_label)  . "=? " . $customSQL .
-      $dateSql, array($data,$patient_id) );
+      if ($whereTables=="" && strpos($table, 'form_')!== false){
```
Do you also need to support when a column is not given?

eyalvo-matrix replied:
No , this handles only the specific case that a column is given 
and  that the table represent a form

bradymiller commented on library/clinical_rules.php:
```diff
@@ -1914,13 +1914,28 @@ function exist_database_item($patient_id,$table,$column='',$data_comp,$data='',$
+              "WHERE a.deleted != '1'".
+              "AND b.`" .add_escape_custom($column) ."`" . $compSql . "? " .
+              "AND b."  . add_escape_custom($patient_id_label)  . "=? " . $customSQL
+              .substr_replace($dateSql,'b.',5,0)
```
Was hoping to get clarification on what this substr_replace of the $dateSql is doing.

eyalvo-matrix replied:
The substr_replace intended to change the $dateSql
 from 
AND (`date` BETWEEN DATE_SUB('......
to  
AND (b.`date` BETWEEN DATE_SUB('......
adding "b." to date 
I did that here and not in the creation of  $dateSql in order to leave the else case unchanged

eyalvo-matrix replied:
Before the change the $dateSql contained something like this :
AND (`date` BETWEEN DATE_SUB('2016-12-28 11:53:37', INTERVAL 6 MONTH) AND '2016-12-28 11:53:37')

bradymiller replied:
thanks for the explanation. makes good sense.

bradymiller commented:
Posted some questions/comments in the code.(minor issues)
thanks, -brady

mdsupport commented:
@eyalvo-matrix,
exist_database_item seemed to be scouring the database for all patient data.  Since there are existing variations in encounter ->> forms access, I want to suggest making PR 359 type of approach reusable and then build the list of exceptions rather than relying on hard coded table names put in by other developers.
Cheers.

eyalvo-matrix commented:
Ok , I will review the PR 359 in depth and try to re-post the code here . 
It will take few days. If anyone else wants to take the gloves feel free :)

bradymiller commented:
@eyalvo-matrix and @mdsupport ,

I'd consider using mdsupport's proposed approach optional (it would, however, be a extremely helpful feature in the long run).

Since your goal was to simply not include deleted forms, I think your current code submission would be acceptable to go in the codebase as long as it tests well (just please ensure that the rules that test the vitals signs are working with your changes).

eyalvo-matrix commented:
Hi Brady I have applied the fixes can you please review the code ?
Thanks a lot ! Eyal .

bradymiller commented:
Hi @eyalvo-matrix ,

Do you mean this PR or your other PR here:
https://github.com/openemr/openemr/pull/373

For this PR, I just need to know that your testing with the vitals rules are still working with your changes (and then I'd bring this PR in)

For the other PR, note my comment on changing the prices_history table to something like codes_history table.

thanks, -brady

eyalvo-matrix commented:
Hi

As far as I checked For this PR - vitals rules indeed still working though
I do not know how to write and check complex rules

About the #373 <https://github.com/openemr/openemr/pull/373> I refactored
the code and changed it to code history as you instructed

about the reports we indeed planning (Not official yet ) to integrate
Openemr with Pentaho BI .

On Mon, Dec 5, 2016 at 10:16 AM, Brady Miller <notifications@github.com>
wrote:

> Hi @eyalvo-matrix <https://github.com/eyalvo-matrix> ,
>
> Do you mean this PR or your other PR here:
> #373 <https://github.com/openemr/openemr/pull/373>
>
> For this PR, I just need to know that your testing with the vitals rules
> are still working with your changes (and then I'd bring this PR in)
>
> For the other PR, note my comment on changing the prices_history table to
> something like codes_history table.
>
> thanks, -brady
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openemr/openemr/pull/360#issuecomment-264791904>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ATqohLN2eyt0H9QKAqwoULD5MB7NMPGMks5rE8hOgaJpZM4K9LPB>
> .
>

bradymiller commented:
just brought this into the codebase. thanks for the fix!

eyalvo-matrix commented:
Thanks :)

bradymiller commented:
FYI,

There were reports of the following error happening (not on all systems):

bradymiller commented:
oops, will try again:

FYI,

There were reports of the following error happening (not on all systems):
![](https://files.slack.com/files-pri/T1MHME2K1-F3D726URK/pasted_image_at_2016_12_10_12_23_am.png)

I committed a fix for this to the codebase:
https://github.com/openemr/openemr/commit/19664124ee7fe8203fd6f3989d8778214728fe56

bradymiller commented:
To be clear, I committed a fix for this here:
https://github.com/openemr/openemr/commit/19664124ee7fe8203fd6f3989d8778214728fe56

eyalvo-matrix commented:
thanks :)
